### DISCUSSION TITLE
psycopg: overridden dialect function takes effect with sync but not with async

### DISCUSSION THREAD

**Discussion by gordthompson** [sqlalchemy/sqlalchemy#9452] [Usage Questions] [answered]
I'm working on adding support for psycopg (v3) to the sqlalchemy-cockroachdb dialect. I've gotten sync code to work, but have hit a snag with async.

This works with asyncpg:

```python
import asyncio

from sqlalchemy.ext.asyncio import create_async_engine

async def async_main():
    connection_url = "cockroachdb+asyncpg://root@localhost:26257/defaultdb"
    # connection_url = "cockroachdb+psycopg://root@localhost:26257/defaultdb"
    engine = create_async_engine(connection_url)
    async with engine.begin() as conn:
        resultset = await conn.exec_driver_sql("SELECT 1 AS foo")
        print(resultset.scalar())

asyncio.run(async_main())
```

However, if I use the `cockroachdb+psycopg` URL I get an error message from `_get_server_version_info()` in SQLAlchemy's own PGDialect:

[stack_trace_psycopg.txt](https://github.com/sqlalchemy/sqlalchemy/files/10923632/stack_trace_psycopg.txt)

What's puzzling me is that `class CockroachDBDialect(PGDialect)` overrides `_get_server_version_info()`

https://github.com/gordthompson/sqlalchemy-cockroachdb/blob/9a52215c142b9b6cc10b5316bbd3d0a6a8b6443b/sqlalchemy_cockroachdb/base.py#L136

and that override takes effect for asyncpg but not for psycopg. For comparison, if I add an `assert False` to the overriden `_get_server_version_info()` and run the code with asyncpg the stack trace looks like this:

[stack_trace_asyncpg.txt](https://github.com/sqlalchemy/sqlalchemy/files/10923639/stack_trace_asyncpg.txt)

And, as I said, I have no problems when I use `create_engine()` with psycopg (i.e., the override works for both psycopg and psycopg2). It's just when I use `create_async_engine()` with psycopg that the override is ignored.

Am I missing something obvious?

**Comment by zzzeek** [comment]
I think you have to include subclasses of the specific dialects also, in this case ``PGDialect_psycopg``.  Then you'd have an entrypoint ``cockroachdb+psycopg`` that returns the ``CockroachDBDialect_psycopg(PGDialect_pyscopg)`` implementation.

**Reply by zzzeek** [reply]
put ``breakpoint()``  commands into the constructors of these dialects (the `__init__()` method).  See where it's getting called (likely in url.get_dialect()).  that will show where you need to make sure your dialect is called.

**Comment by CaselIT** [comment, marked as answer]
Hey gord.
It's probably the fact that there is an adapter to allow psycopg to be both sync and async.

Ensure that you overide this
https://github.com/sqlalchemy/sqlalchemy/blob/8d5986fafd8360ddfe3992bd56602d5a52a23392/lib/sqlalchemy/dialects/postgresql/psycopg.py#L374-L376
so that the enigine resolves the correct version of the dialect, otherwise you will get the default psycopg_async one

**Reply by gordthompson** [reply]
That helped. Thanks!
